WalletWasabi.Backend/Controllers/BlockchainController.cs

```diff
@@ -416,4 +416,88 @@ private async Task<StatusResponse> FetchStatusAsync(CancellationToken cancellati
 
 		return status;
 	}
+
+	[HttpGet("unconfirmed-transaction-chain")]
+	[ProducesResponseType(200)]
+	[ProducesResponseType(400)]
+	public async Task<List<UnconfirmedTransactionChainItem>> GetUnconfirmedTransactionChainAsync([FromQuery, Required] string transactionId, CancellationToken cancellationToken)
+	{
+		uint256 txId = new(transactionId);
+
+		var cacheKey = $"{nameof(GetUnconfirmedTransactionChainAsync)}_{txId}";
+		var cacheOptions = new MemoryCacheEntryOptions { AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(3) };
+
+		return await Cache.GetCachedResponseAsync(
+			cacheKey,
+			action: (string request, CancellationToken token) => GetUnconfirmedTransactionChainNoCacheAsync(txId, token),
+			options: cacheOptions,
+			cancellationToken);
+	}
+
+	private async Task<List<UnconfirmedTransactionChainItem>> GetUnconfirmedTransactionChainNoCacheAsync(uint256 txId, CancellationToken cancellationToken)
+	{
+		Dictionary<uint256, Transaction> transactionsLocalCache = new();
+		Dictionary<uint256, UnconfirmedTransactionChainItem> unconfirmedTxsChainById = new();
+		var mempool = Global.HostedServices.Get<MempoolMirror>();
+
+		// TODO: Use Transaction cache.
+		var requestedTransaction = await RpcClient.GetRawTransactionAsync(txId, true, cancellationToken);
+
+		List<Transaction> toFetchFeeList = new() { requestedTransaction };
+
+		while (toFetchFeeList.Count > 0)
+		{
+			List<Coin> inputs = new();
+			HashSet<Transaction> parentTxs = new();
+
+			var currentTx = toFetchFeeList.First();
+			foreach (var input in currentTx.Inputs)
+			{
+				if (!transactionsLocalCache.TryGetValue(input.PrevOut.Hash, out var parentTx))
+				{
+					parentTx = await RpcClient.GetRawTransactionAsync(input.PrevOut.Hash, true, cancellationToken);
+					transactionsLocalCache.Add(input.PrevOut.Hash, parentTx);
+				}
+
+				parentTxs.Add(parentTx);
+				TxOut txOut = parentTx.Outputs[input.PrevOut.N];
+				inputs.Add(new Coin(input.PrevOut, txOut));
+			}
+
+			var unconfirmedChildrenTxs = mempool
+				.GetSpenderTransactions(currentTx.Outputs.Select((txo, index) => new OutPoint(currentTx, index)))
+				.ToHashSet();
+
+			// Add the children to the local transaction cache.
+			foreach (var childTx in unconfirmedChildrenTxs)
+			{
+				if (!transactionsLocalCache.ContainsKey(childTx.GetHash()))
+				{
+					transactionsLocalCache.Add(childTx.GetHash(), childTx);
+				}
+			}
+
+			// Remove the item we worked on.
+			toFetchFeeList.Remove(currentTx);
+
+			var unconfirmedParents = parentTxs.Where(x =>
+				mempool.GetMempoolHashes().Contains(x.GetHash()));
```

Comment:
Store the mempool hashes at the beginning of this method and use them during the entire call. To ensure the integrity of what we are doing. So basically take a snapshot of the mempool. Also getting the mempool is locking so better to do it only once per API call.